### PR TITLE
Use multiple IovArray for writev when using io_uring based transport

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

How we did manage the memory of writev was quite wasteful and could
produce a lot of memory overhead. We can just keep it simple by using
one IovArray. Once it is full we can just submit and clear it as at this
point the kernel did take over a copy and its safe to reuse

Modifications:

Use one IovArray and submit once it is full.

Result:

Less memory overhead and less code duplication

### CODE REVIEW

File: transport-native-io_uring/src/main/java/io/netty/channel/uring/AbstractIOUringChannel.java [hunk 1]
```diff
@@ -304,23 +302,17 @@ protected void doWrite(ChannelOutboundBuffer in) {
...
+         final IovArray iovecArray = ((IOUringEventLoop) eventLoop()).iovArray();
+         try {
+             int offset = iovecArray.count();
+             in.forEachFlushedMessage(iovecArray);
+             submissionQueue().addWritev(socket.intValue(), iovecArray.memoryAddress(offset), iovecArray.count() - offset);
              ioState |= WRITE_SCHEDULED;
-         } else {
-             // We were not be able to create a new iovec, fallback to single write.
+         } catch (Exception e) {
+             // This should never happem, anyway fallback to single write.
```
Comment (chrisvest) [inline transport-native-io_uring/src/main/java/io/netty/channel/uring/AbstractIOUringChannel.java:312]: `happem`

**Review by chrisvest — APPROVED** [review]
One nit comment, but pre-approved.

File: transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringEventLoop.java [hunk 2]
```diff
@@ -268,7 +278,12 @@ protected void wakeup(boolean inEventLoop) {
         }
     }
 
-    public IovecArrayPool getIovecArrayPool() {
-        return iovecArrayPool;
+    public IovArray iovArray() {
+        // Check if its full and if so submit so we can reuse it.
+        if (iovArray.isFull()) {
```
Comment (1Jo1) [inline transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringEventLoop.java:283]: one drawback would be that we reduce submission batch size when a lot of channel writing in the same time especially with a bigger message size, one solution would be: to make the capacity configurable, default iovec array size is only 1024 as far as I know

**Review by 1Jo1 — CHANGES_REQUESTED** [review]
looks good, just one comment

File: transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringEventLoop.java (hunk 2 above)
Comment (normanmaurer) [reply transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringEventLoop.java:283]: yep I am aware of this... thats why I added some "we may want to adjust this later comment"... That said let me run some benchmarks for now first

**Comment by normanmaurer** [comment]
@1Jo1 @chrisvest I did revisit this and decided we should just use 8 `IovArray` instances per `EventLoop`. This still gives plenty of room while not being crazy in terms of memory overhead.  PTAL again

File: transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringEventLoop.java [hunk 3]
```diff
@@ -46,18 +47,26 @@
...
     private long prevDeadlineNanos = NONE;
     private boolean pendingWakeup;
 
     IOUringEventLoop(final EventLoopGroup parent, final Executor executor, final boolean addTaskWakesUp) {
         super(parent, executor, addTaskWakesUp);
+        // Ensure that we load all native bits as otherwise it may fail when try to use native methods in IovArray
+        IOUring.ensureAvailability();
 
+        // Let's hard code this to 8 IovArrays to keep the memory overhead kind of small.
+        iovArrays = new IovArray[8];
```
Comment (normanmaurer) [inline transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringEventLoop.java:63]: we could even make this more optimal by writing a custom `IovArray` implementation. But for now this seems to work well enough and allows us for code-reuse.

**Review by chrisvest — APPROVED** [review]

**Comment by 1Jo1** [comment]
@normanmaurer btw IOUringSocketGatheringWriteTest seems to be very slow..

**Comment by normanmaurer** [comment]
Will check ... for the benchmark it did not have any negative effect

**Comment by njhill** [comment]
@normanmaurer how about we just allocate a `ByteBuf` as needed?

**Review by 1Jo1 — APPROVED** [review]
it looks good to me :+1: did some benchmark tests, 8 iovec arrays should be enough

File: transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringEventLoop.java (hunk 3 above)
Comment (1Jo1) [reply transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringEventLoop.java:64]: I like that, 8 we should take it as default value when it's configurable

**Review by chrisvest — APPROVED** [review]

**Comment by normanmaurer** [comment]
@njhill I think for now its ok like it is... let me merge this and we can revisit later if needed.

**Comment by njhill** [comment]
@normanmaurer I just realized that there can also be just-in-time `submit()` inside `IOUringSubmissionQueue` when adding SQE to full queue. Probably would be best to clear the `iovArrays` after those too?

Maybe we could have a post-submission callback to the EL?

**Comment by normanmaurer** [comment]
@njhill good point. I will do a followup pr tomorrow once I wake up :)

**Pull request merged**